relayer/chains/substrate/event_parser.go

```diff
@@ -1,29 +1,501 @@
 package substrate
 
 import (
+	"fmt"
+	"strconv"
+
 	rpcclienttypes "github.com/ComposableFi/go-substrate-rpc-client/v4/types"
+	beefyclienttypes "github.com/ComposableFi/ics11-beefy/types"
 	clienttypes "github.com/cosmos/ibc-go/v5/modules/core/02-client/types"
-	"github.com/cosmos/relayer/v2/relayer/processor"
 	"github.com/cosmos/relayer/v2/relayer/provider"
+	"go.uber.org/zap"
+	"go.uber.org/zap/zapcore"
 )
 
-// ibcMessagesFromTransaction parses all events within a transaction to find IBC messages
-func (scp *SubstrateChainProcessor) handleIBCMessagesFromEvents(ibcEvents rpcclienttypes.IBCEventsQueryResult, height uint64, c processor.IBCMessagesCache) error {
-	//TODO implement me
-	panic("implement me")
+// ibcMessage is the type used for parsing all possible properties of IBC messages
+type ibcMessage struct {
+	eventType string
+	info      ibcMessageInfo
+}
+
+type ibcMessageInfo interface {
+	parseAttrs(log *zap.Logger, attrs interface{})
+	MarshalLogObject(enc zapcore.ObjectEncoder) error
+}
+
+// alias for the interface map
+type ibcEventQueryItem (map[string]interface{})
+
+// substrate ibc events endpoint returns a list of events
+// relayted to different transaction, so we need
+// to define a unique key
+type ibcPacketKey struct {
+	sequence   uint64
+	srcChannel string
+	srcPort    string
+	dstChannel string
+	dstPort    string
+}
+
+// ibcMessagesFromEvents parses all events of a certain height to find IBC messages
+func (scp *SubstrateChainProcessor) ibcMessagesFromEvents(
+	ibcEvents rpcclienttypes.IBCEventsQueryResult,
+	height uint64,
+) (messages []ibcMessage) {
+
+	packetAccumulator := make(map[ibcPacketKey]*packetInfo)
+	for i := 0; i < len(ibcEvents); i++ {
+
+		info, eventType := scp.parseEvent(ibcEvents[i], height, packetAccumulator)
+		if info == nil {
+			// Not an IBC message, don't need to log here
+			continue
+		}
+
+		// update messages
+		messages = append(messages, ibcMessage{
+			eventType: eventType,
+			info:      info,
+		})
+	}
+
```

Comment:
here we should iterate over `packetAccumulator` and append to `messages` for each with an `ibcMessage` like:
```golang
ibcMessage {
  eventType: chantypes.EventTypeRecvPacket,
  info: pi,
}
```